### PR TITLE
Experimental - unit-testing of backend functions

### PR DESCRIPTION
**Pull request by xmano**
Changes for Unit Test infra

### CODE REVIEW

**Comment by xmano** [comment]
Got all tests to work correctly.. need to cleanup

image

**Comment by xmano** [comment]
Need to fix an issue with  UTs

https://app.circleci.com/pipelines/github/blockframes/blockframes/22184/workflows/6b8ee5f7-f8be-4d1f-8561-933566e56784

**Comment by xmano** [comment]
Have resolved all open issues with the PR and fixed the TODOs. Ready for review

image

File: apps/backend-functions/src/fixtures/data.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+export const testFixture = {
+  //Meta
+  '_META/_MAINTENANCE': {
+    startedAt: null,
+    endedAt: null
+  },
+
+  '_META/_VERSION': {
+    currentVersion: 69
```
Comment (phpgeekfr) [inline apps/backend-functions/src/fixtures/data.ts:11]: do we need this ? as currentVersion will change

File: apps/backend-functions/src/invitation.spec.ts [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    await initFirestoreApp(projectRealId, '', testFixture, { uid: 'uid-user2', firebase: { sign_in_provider: 'password' } });
+  });
+
+  afterAll(async () => {
+    // After each test, db is reseted
+    await clearFirestoreData({ projectId: projectRealId });
+  });
+
+
+  describe("Invitation spec", () => {
```
Comment (phpgeekfr) [inline apps/backend-functions/src/invitation.spec.ts:32]: use single quotes

File: apps/backend-functions/src/invitation.spec.ts [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        emails: ['test@cascade8.com'],
+        invitation: {},
+        app: 'catalog'
+      };
+
+      expect.assertions(1);
+      await expect(async ()=> {
+        await wrapped(data, { })
+      }).rejects
+        .toThrow("Permission denied: missing auth context.");
```
Comment (phpgeekfr) [inline apps/backend-functions/src/invitation.spec.ts:47]: use single quotes

File: apps/backend-functions/src/invitation.spec.ts [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+          uid: 'uid-c8',
+          token: ''
+        }
+      };
+
+      expect.assertions(1);
+      await expect(async ()=> {
+        await wrapped(data, context)
+      }).rejects
+        .toThrow("Permission denied: missing org id.");
```
Comment (phpgeekfr) [inline apps/backend-functions/src/invitation.spec.ts:71]: use single quotes

File: apps/backend-functions/src/invitation.spec.ts [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+          uid: 'uid-user2',
+          token: ''
+        }
+      };
+
+      const result = await wrapped(data, context);
+      expect(result).toEqual([]);
+    });
+
+    it("For 'Join organization' event, email is sent & invite doc created", async () => {
```
Comment (phpgeekfr) [inline apps/backend-functions/src/invitation.spec.ts:100]: use single quotes

File: apps/backend-functions/src/invitation.spec.ts [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+      //Check results have correct data
+      expect(result.length).toEqual(1);
+      expect(result[0]).toEqual(
+        expect.objectContaining({
+          error: ''
+        })
+      );
+
+      const inviteId = result[0].id;
+      const snap = await admin.firestore().collection("invitations").doc(inviteId).get();
```
Comment (phpgeekfr) [inline apps/backend-functions/src/invitation.spec.ts:148]: use single quotes

File: apps/backend-functions/src/invitation.spec.ts [hunk 7]
```diff
@@ -0,0 +1,158 @@
+﻿import { resolve } from 'path';
+process.env.FIRESTORE_EMULATOR_HOST = 'localhost:8080';
+import { initFirestoreApp } from '@blockframes/testing/firebase/functions';
```
Comment (phpgeekfr) [inline apps/backend-functions/src/invitation.spec.ts:3]: from `'@blockframes/testing/unit-tests';`

Comment (xmano) [reply apps/backend-functions/src/invitation.spec.ts]: Already done.

File: apps/backend-functions/src/invitation.spec.ts (hunk 2 above)
Comment (xmano) [reply apps/backend-functions/src/invitation.spec.ts]: done

File: apps/backend-functions/src/invitation.spec.ts (hunk 4 above)
Comment (xmano) [reply apps/backend-functions/src/invitation.spec.ts]: done

File: apps/backend-functions/src/invitation.spec.ts (hunk 5 above)
Comment (xmano) [reply apps/backend-functions/src/invitation.spec.ts]: done

File: apps/backend-functions/src/invitation.spec.ts (hunk 6 above)
Comment (xmano) [reply apps/backend-functions/src/invitation.spec.ts]: done

File: apps/backend-functions/src/invitation.spec.ts (hunk 3 above)
Comment (xmano) [reply apps/backend-functions/src/invitation.spec.ts]: done

**Comment by xmano** [comment]
Pretty tough conflict to resolve:

backend-functions work with @firebase/rules-unit-testing library

whereas rules need  @firebase/testing 

Need to resolve these differences..

File: apps/backend-functions/src/fixtures/data.ts (hunk 1 above)
Comment (xmano) [reply apps/backend-functions/src/fixtures/data.ts]: removed this.

**Comment by xmano** [comment]
I created a new library file to overcome the library issue src/internals/firebase-test.ts.

This is a workaround for now.

File: apps/backend-functions/src/invitation.spec.ts [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+const pathToServiceAccountKey = resolve(process.cwd(), process.env.GOOGLE_APPLICATION_CREDENTIALS);
+const testEnv = firebaseTest(firebase(), pathToServiceAccountKey);
+
+describe('Invitation backend-function unit-tests', () => {
+
+  beforeAll(async () => {
+    process.env.FIRESTORE_EMULATOR_HOST = 'localhost:8080';
+    // We are initing firestore with client settings to
+    // test the functions effectively
+    await initFirestoreApp(projectRealId, '', testFixture, { uid: 'uid-user2', firebase: { sign_in_provider: 'password' } });
```
Comment (phpgeekfr) [inline apps/backend-functions/src/invitation.spec.ts:24]: looks like it is working with only `await initFirestoreApp(projectRealId, '', testFixture);`

File: apps/backend-functions/src/invitation.spec.ts [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+import { expect } from '@jest/globals';
+
+const projectRealId = firebase().projectId;
+const pathToServiceAccountKey = resolve(process.cwd(), process.env.GOOGLE_APPLICATION_CREDENTIALS);
+const testEnv = firebaseTest(firebase(), pathToServiceAccountKey);
+
+describe('Invitation backend-function unit-tests', () => {
+
+  beforeAll(async () => {
+    process.env.FIRESTORE_EMULATOR_HOST = 'localhost:8080';
```
Comment (phpgeekfr) [inline apps/backend-functions/src/invitation.spec.ts:21]: duplicated line ..

File: apps/backend-functions/src/internals/firebase.ts [hunk 10]
```diff
@@ -1,13 +1,25 @@
...
 if (!admin.apps.length) {
-  admin.initializeApp();
+  if (emulation) {
+    const projectId = firebaseEnv().projectId;
+    initFunctionsTestMock(true, {projectId});
+    //Setup the maintenance document for real project
+    admin.firestore().doc(`${META_COLLECTION_NAME}/${MAINTENANCE_DOCUMENT_NAME}`).set({
+       startedAt: null,
+       endedAt: new Date(Date.now() - 60 * 60 * 1000)
+    });
```
Comment (phpgeekfr) [inline apps/backend-functions/src/internals/firebase.ts:19]: use endMaintenance directly in the spec.ts file

File: apps/backend-functions/src/internals/firebase.ts [hunk 11]
```diff
@@ -1,13 +1,25 @@
...
 import { backupBucket, storageBucket } from '../environments/environment';
 import { defaultConfig, isInMaintenance } from '@blockframes/firebase-utils';
 import { IMaintenanceDoc, META_COLLECTION_NAME, MAINTENANCE_DOCUMENT_NAME, _isInMaintenance } from '@blockframes/utils/maintenance';
+import { initFunctionsTestMock } from '@blockframes/testing/unit-tests';
+const emulation = (process.env.FIRESTORE_EMULATOR_HOST === 'localhost:8080');
 
 if (!admin.apps.length) {
-  admin.initializeApp();
+  if (emulation) {
+    const projectId = firebaseEnv().projectId;
```
Comment (phpgeekfr) [inline apps/backend-functions/src/internals/firebase.ts:13]: you cannot use `getTestingProjectId()` ?

File: apps/backend-functions/src/invitation.spec.ts (hunk 8 above)
Comment (phpgeekfr) [reply apps/backend-functions/src/invitation.spec.ts:24]: and why not using `initFunctionsTestMock` as other tests ?

File: apps/backend-functions/src/invitation.spec.ts (hunk 9 above)
Comment (phpgeekfr) [reply apps/backend-functions/src/invitation.spec.ts:21]: and also in `initFirestoreApp`

**Review by phpgeekfr — APPROVED** [review]

**Pull request merged**